### ISSUE TITLE
[Ruby Rails] URL rewrite clips path (embedded in rails app)

### ISSUE THREAD

**Issue by Kharonus** [Rhosys/openapi-explorer#202]
First of all, thanks a lot for this OpenAPI explorer. We had in bunch in research, when migrating to OpenAPI 3.1, and this one was the clear winner in terms of flexibility in use and completeness of functionality.

The headline describes my problem as a wild guess, as I couldn't prove it to be the real reason. Here is what happens:

### Issue

- The `openapi-explorer` is rendered at the route `<HOST>/api/docs`
- accessing this URL opens the explorer and loads the view
- after this is finished, the URL is rewritten to `<HOST>/#?route=overview`
  - note the missing URL path, making the URL invalid for a refresh or a copy/share
  - the explorer itself works flawlessly, as long as no one refreshes the page

### Some details to the tech stack

- it is embedded in the real application, i.e. the menu of the app is still visible on top
- the app is a rails application, we wrote a controller, that is responsible for rendering the explorer
- the controller's view is as simple as

```erb
<%= javascript_include_tag "openapi-explorer.min.js" %>

<%= content_tag('openapi-explorer',
                '',
                'spec-url': api_v3_paths.api_spec,
                'nav-bg-color': color_by_variable('main-menu-bg-color')) %>
```

- the `content_tag` helper is a mere ruby helper, that creates the the html tag `<openapi-explorer></openapi-explorer>` with empty content and with the attributes listed in there

### Question and problem finding

I know, that this information is very minimal. I myself have a hard time to figure the correct approach and gather more information.

Is anything like this known to you or does the problem with the missing URL path rings any bells? I'm grateful for any clue, that can help me.

BR
kharonus

**Comment by wparad** [comment]
The rewrite happens whenever the "view" of the component changes. The code called is in here: https://github.com/Rhosys/openapi-explorer/blob/release/2.1/src/utils/common-utils.js#L140

```js
export function replaceState(rawElementId) {
  const elementId = rawElementId && rawElementId.replace(/^#/, '') || '';

  const currentNavigationHashPart = (window.location.hash || '').split('?')[0].replace(/^#/, '');
  const currentQuery = (window.location.hash || '').split('?')[1];
  const query = new URLSearchParams(currentQuery);
  query.delete('route');
  const newQuery = query.toString().length > 1 ? `${query.toString()}&route=${elementId}` : `route=${elementId}`;
  window.history.replaceState(null, null, `#${currentNavigationHashPart}?${newQuery}`);
}
```

Taking another look at this, there is nothing here that should cause a problem, and you can actually see in the production version of Authress, we too have actually both `path` segments and `hash` segments, and this still works correctly:
https://authress.io/app/#/api?route=get-/v1/users

So I'm at a bit of a loss what could be happening here. My suggestion would be to check the logs from rails and see what the request/responses are. When the path in the UI gets rewritten it should become `<HOST>/api/docs/#?route=overview`, but only flash for a moment on the screen. In this case you should see a request to rails with `<HOST>/api/docs/#?route=overview` but then a response that incorrectly changes the url.

However, looking at the code it explicitly only updates the `hash` part, and not anything more than that. Which has to affects:
* No request should be sent to rails
* The rest of the path works correctly

Now, it's certainly possible that this is a browser issue and not a rails ones as well. I've just tested `window.history.replaceState` on firefox, chrome, and brave. Are you using a different browser/os here?

**Comment by Kharonus** [comment]
Thanks for the very quick answer!

I was testing this on brave, and `replaceState` works as expected. I will now follow the leads you provided and try to debug both rails logs and the js snippet you pointed out.

**Comment by Kharonus** [comment]
I verified, that `openapi-explorer` behaves as expected, and the issue lies on our side in a weird conflict with the UI router of our JS framework (which we should have removed long ago... damn...).

Hence, this is no issue that can be solved on your side. But big thanks to your troubleshooting leads! You can close that issue, or leave it open for a while. If I detect the conflicting source, I can share it as an experience for things, that can cause trouble.

Best,
kharonus

**Comment by wparad** [comment]
:+1: Glad you at least figured out the source of the issue. I'm going to close this, if we figure out a potential solution we can still add it in. My only concern would be I don't know if there is actually a way for the library to test if this sort of behavior will happen or not, and we can't change this code to always put the path in there because that will in some cases force a page reload. I think the code is as correct as it could be, but I'm also not a browser expert when it comes to things like this.

-----

Some of my testing:

Wow, yeah, so I've tested the `window.history.replaceState(null, null, '#new?path=true')` on a few different websites and they all work correctly. So something is definitely wrong with the `openproject.org` (at least the login page).

The problem is I can't narrow down why this happening. I've removed javascript execution from the site, and it still happens. I also validated that there are no network requests happening.

**Comment by Kharonus** [comment]
Yeah, I'm also glad we found that issue. And parts of the website are public and accessible without login, like https://community.openproject.org/work_packages, but it happens all the same.

I agree with your code being as correct as it could be. More important for me even, it helped me finding this issue on our app. ;)

thanks for your time!

**Comment by Kharonus** [comment]
@wparad I may have an interesting fact for you:

the source of this behavior might(!) be the angular ui router. at least, one can observe the same behavior on other angular apps. (e.g. here: https://angular.io/guide/developer-guide-overview).

Strangely enough, I couldn't find an issue yet in the [angular repo](https://github.com/angular/angular/issues) reporting it.

**Comment by wparad** [comment]
I'm guessing it is related to this SO Post: [https://stackoverflow.com/questions/37927558/remove-hashchange-event-from-angular-or-prevent-angular-from-rewriting-anchor-](Remove hashchange event from angular) 

There's also this one, not sure how "related" it is:
* https://github.com/angular/angular/issues/2555
* https://github.com/angular/angular/issues/28954
* https://github.com/angular/angular/issues/41782

I'm assuming it has to do with an incorrect implementation of the hash router. https://github.com/angular/angular/blob/327896606832bf6fbfc8f23989755123028136a8/packages/common/src/location/platform_location.ts#L132

You can see that the hashChange function expects that there to be a full URL: https://github.com/angular/angular/blob/327896606832bf6fbfc8f23989755123028136a8/packages/platform-server/src/location.ts#L127 I'm guessing somewhere, they forgot to pass the full thing, I can't find full stack for where this works though.

**Comment by supremedictator** [comment]
I came across this issue after having a similar issue and thought I would add my findings incase someone else finds it useful.

Background:
My application is an ASP .Net 8 app service using Angular 19.
I added this openapi-explorer using npm and followed the instructions for the vanilla html usecase. I also setup my angular component so it would work with the web component.

Everything loaded and worked well except I was having the same issue as described here.
My URL was localhost:4444/documentation. However clicking a link (and on initial load of the documentation) it would change to localhost:4444/#?route=overview for example.

After a few hours of investigating I was able to find the cause for me was the <base href="/"> tag that was in my index.html file.
This was automatically added by the templating in VS2022.

When the hash routing was pushed to the browser history it appears it was doing so relative to that tag.
Removing that tag seems to make it relative to the current url and path and stops the path being stripped
